include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+  Evaluation(const Evaluation &) = default;
+
+  /// General ctor
+  template <typename A>
+  Evaluation(const A &a, const parser::CharBlock &pos,
+             const std::optional<parser::Label> &lab, const ParentType &p)
+      : u{&a}, parent{p}, pos{pos}, lab{lab} {}
+
+  /// Compiler-generated jump
+  Evaluation(const CGJump &jump, const ParentType &p)
+      : u{jump}, parent{p}, cfg{CFGAnnotation::Goto} {}
+
+  /// Construct ctor
+  template <typename A>
+  Evaluation(const A &a, const ParentType &parent) : u{&a}, parent{parent} {
+    static_assert(AST::isConstruct<A>(), "must be a construct");
+  }
+
+  /// is `A` executable (an action statement or compiler generated)?
+  template <typename A>
+  constexpr static bool isAction(const A &a) {
+    return !AST::isConstruct<A>() && !isOther(a);
+  }
+
+  /// is `A` a compiler-generated evaluation?
+  template <typename A>
+  constexpr static bool isGenerated(const A &) {
+    return std::is_same_v<A, CGJump>;
+  }
+
+  /// is `A` not an executable statement?
+  template <typename A>
+  constexpr static bool isOther(const A &) {
+    return std::is_same_v<A, parser::FormatStmt> ||
+           std::is_same_v<A, parser::EntryStmt> ||
+           std::is_same_v<A, parser::DataStmt> ||
+           std::is_same_v<A, parser::NamelistStmt>;
+  }
+
+  constexpr bool isActionStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+
+  constexpr bool isStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p) || isOther(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+  constexpr bool isConstruct() const { return !isStmt(); }
+
+  /// Set the type of originating control flow type for this evaluation.
+  void setCFG(CFGAnnotation a, Evaluation *cstr) {
+    cfg = a;
+    setBranches(cstr);
+  }
+
+  /// Is this evaluation a control-flow origin? (The AST must be annotated)
+  bool isControlOrigin() const { return cfg != CFGAnnotation::None; }
+
+  /// Is this evaluation a control-flow target? (The AST must be annotated)
+  bool isControlTarget() const { return isTarget; }
+
+  /// Set the containsBranches flag iff this evaluation (a construct) contains
+  /// control flow
+  void setBranches() { containsBranches = true; }
+
+  constexpr std::list<Evaluation> *getConstructEvals() {
+    return isStmt() ? nullptr : subs;
+  }
+
+  /// Set that the construct `cstr` (if not a nullptr) has branches.
+  static void setBranches(Evaluation *cstr) {
+    if (cstr) {
+      cstr->setBranches();
+    }
+  }
+
+  EvalVariant u;
+  ParentType parent;
+  parser::CharBlock pos;
+  std::optional<parser::Label> lab;
+  std::list<Evaluation> *subs{nullptr}; // construct sub-statements
```

Comment:
Most Evaluations do not contain a sub-list. The pointer omits calling the std::list constructor as tax on every node in the AST.

Reply:
Is a unique_ptr suitable here?

Reply:
The constructors for `std::list` and `std::vector` are both empty (and trivially inlineable as they are templates), at least in both `libc++` and `libstdc++`. This is so that people can use empty containers in cases like this without issues.

Reply:
@DavidTruby Thanks for pointing that out.

Logically, most `Evaluations` have no body. It makes no sense. One solution would be to use a subclass here. Another would be to use `std::optional<>`. The pointer-to solution here is, effectively, the `std::optional` solution used to work around other issues. The least desirable option, in my opinion, would be to have an list everywhere even if the run-time cost is zero since it would entail two distinct logical states. (1) This `Evaluation` contains other `Evaluation`s but not in this instance. (2) This `Evaluation` can absolutely, positively never contain other `Evaluation`s.

Reply:
Updated to a `std::unique_ptr`. It keeps the the state indication mentioned by @schweitzpgi (different form an empty list).